SerialPrograms/Source/CommonFramework/AudioPipeline/AudioDisplayWidget.cpp

```diff
@@ -0,0 +1,234 @@
+/*  Audio Display Widget
+ *
+ *  From: https://github.com/PokemonAutomation/Arduino-Source
+ *
+ */
+
+#include "AudioDisplayWidget.h"
+#include "AudioInfo.h"
+#include "AudioThreadController.h"
+#include "CommonFramework/Logging/Logger.h"
+#include "Kernels/AbsFFT/Kernels_AbsFFT.h"
+
+#include <QVBoxLayout>
+#include <QLabel>
+#include <QLinearGradient>
+#include <QResizeEvent>
+#include <QPaintEvent>
+#include <QPainter>
+#include <QColor>
+#include <QIODevice>
+#include <QBuffer>
+
+#include <iostream>
+#include <cfloat>
+
+#ifdef USE_FFTREAL
+#include <fftreal_wrapper.h>
+#else
+const int FFTLengthPowerOfTwo = 14;
+#endif
+
+const int NUM_FFT_SAMPLES = 1 << FFTLengthPowerOfTwo;
+
+namespace PokemonAutomation{
+
+
+AudioDisplayWidget::AudioDisplayWidget(QWidget& parent)
+     : QWidget(&parent)
+     , m_numFreqs(NUM_FFT_SAMPLES/2-1)
+     , m_numFreqWindows(80)
+     , m_numFreqVisBlocks(32)
+     , m_freqMags(m_numFreqs * m_numFreqWindows)
+     , m_freqVisBlocks(m_numFreqVisBlocks * m_numFreqWindows){}
+
+AudioDisplayWidget::~AudioDisplayWidget(){ clear(); }
+
+void AudioDisplayWidget::clear(){
+    if (m_audioThreadController){
+        delete m_audioThreadController;
+        m_audioThreadController = nullptr;
+    }
+
+    m_freqVisBlocks.assign(m_freqVisBlocks.size(), 0.f);
+    m_freqMags.assign(m_freqMags.size(), 0.f);
+}
+
+void AudioDisplayWidget::close_audio(){
+    clear();
+
+    update_size();
+}
+
+void AudioDisplayWidget::set_audio(Logger& logger, const AudioInfo& inputInfo, const AudioInfo& outputInfo){
+    clear();
+
+    m_audioThreadController = new AudioThreadController(this, inputInfo, outputInfo);
+
+    update_size();
+}
+
+void AudioDisplayWidget::loadFFTOutput(const QVector<float>& fftOutput){

```

Comment:
Is this running on the UI thread?

Reply:
This loadFFTOutput is run on the UI thread. I connected the signals from the FFT thread to this slot. Qt's connection will make sure the slot function is run on the target object's thread, where the target object is from when the connection is defined: connect(source_object, source_signal, target_object, target_slot).